### PR TITLE
Closes #8654: [MCP] add cache status and health abilities

### LINKED ISSUE #8654: [MCP] Check cache status
### What we’re looking to do

This capability is based on three tools that interact with the `wp_wpr_rocket_cache` table:

1. **Page cache lookup**: checks a specific URL's row and returns its status (`pending` / `in-progress` / `completed` / `failed`) plus `modified` (last status change) and `last_accessed`.
2. **Cache health check**: returns counts across all four statuses, plus meta data to help the LLM estimate the estimated completion time.
3. **Cache clear**: wraps WP Rocket's existing `rocket_clean_files()` if using a specific URL or multiple URLs or `rocket_clean_post()` if using a post ID and `rocket_clean_domain()`. 

### Acceptance criteria

- Users can check whether a specific page is cached, and see its current status (pending / in-progress / completed / failed).
- The page-level result includes `modified` (when the status last changed) and `last_accessed`.
- Users can check cache health across the system.
- Cache health returns counts for all four statuses: pending, in-progress, completed, failed.
- Cache health returns an estimated completion time, clearly labeled as an estimate (based on remaining URLs pending URLs, communicated batch size, and if separate cache for mobile is active or not).
- Users can clear the cache for a specific page, using `rocket_clean_files()` or `rocket_clean_post()`.
- Users can clear the cache sitewide, using `rocket_clean_domain()`.
- After a clear is triggered, the user gets confirmation the action was accepted. Regeneration is triggered through the respective above actions.
- A user can trigger a clear even if no problem was detected.

For test

- [ ]  Check cache status of single page - post - custom post - term
- Users can check whether a specific page is cached, and see its current status (pending / in-progress / completed / failed).
- The page-level result includes `modified` (when the status last changed) and `last_accessed`.
- [ ]  Check cache status of whole site
- Cache health returns counts for all four statuses: pending, in-progress, completed, failed.
- Cache health returns an estimated completion time, clearly labeled as an estimate (based on queue depth, not a stored field).
- [ ]  Check clear cache of single page
- After a clear is triggered, the user gets confirmation the action was accepted. Regeneration is asynchronous and processed in batches, so this doesn't need to block until completion.
- [ ]  Check clear cache of whole site
- After a clear is triggered, the user gets confirmation the action was accepted. Regeneration is asynchronous and processed in batches, so this doesn't need to block until completion.
- [ ]  Check clear cache , will clear CF cache for single page or all site
- [ ]  Explore how it works with multilingual
- [ ]  Explore how it works with multisite (subdirectory, subdomain, qs)

### PR DESCRIPTION
**Pull request by wordpressfan**
> 🤖 AI-generated — created by an automated pipeline. Review before acting on this.

Closes #8654

## Description

Adds two new MCP/Abilities tools to expose WP Rocket's Preload cache queue to LLM clients: per-URL/post/term cache status lookups, and a sitewide health summary with ETAs.

**Update:** the originally-included `wp-rocket/clear-cache` ability (scope=url/post/domain) has been removed from this PR. Its `scope=domain` duplicated #8623's `wp-rocket/clear-website-cache` (already merged) and its `scope=url` duplicated #8598's `wp-rocket/clear-url-cache` (same validation, homepage special-case, and `rocket_clean_files()` call) — landing all three would have left three overlapping cache-clearing abilities in the `wp-rocket-cache` category. Post-level cache clearing via MCP can be proposed separately if still needed.

## Type of change

- [x] New feature (non-breaking change which adds functionality).
- [ ] Bug fix (non-breaking change which fixes an issue).
- [ ] Enhancement (non-breaking change which improves an existing functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as before).
- [ ] Sub-task of #8654
- [ ] Chore
- [ ] Release

### What was tested

QA validated this PR against WP 7.0.2 using WP-CLI `wp eval` calls against a live `wp-rocket/*` ability registry plus the automated unit/integration suites. Acceptance criteria covering the two remaining abilities passed:

- `get-cache-status`: live lookup of a tracked URL matched the DB row (`status`, `modified`, `last_accessed` ISO 8601); an untracked URL correctly returned `tracked: false` with an explanatory `note`; multi-identifier and invalid-identifier inputs returned structured `error` values.
- `get-cache-health`: counts matched the seeded table, `tracking_enabled: true`, `estimate.is_estimate: true`, formula documented in `method`.
- Capability enforcement verified live: `rocket_manage_options` required for both read-only abilities; WP < 6.9 no-op verified by code inspection (`function_exists( 'wp_register_ability' )` guard).

Full QA report: https://github.com/wp-media/wp-rocket/pull/8665#issuecomment-5141983780

A follow-up commit removed the `wp-rocket/clear-cache` ability entirely (see Description) after review found it duplicated cache-clearing functionality already present in #8623 (merged) and proposed in #8598 (open).

### How to test

1. **Enable Preload tracking**: Ensure "Preload Cache" is enabled in Settings > WP Rocket > Preload and at least one URL has reached `completed` status (or seeded via `wp_wpr_rocket_cache` table).
2. **Test get-cache-status ability**: Call `wp-rocket/get-cache-status` with `{ "url": "<complete URL>" }` and verify `tracked`, `status`, `modified`, and `last_accessed` match the DB row; test with untracked URLs and confirm `tracked: false` with explanatory note.
3. **Test get-cache-health ability**: Call `wp-rocket/get-cache-health` with no input and verify counts match row counts in the database; confirm `estimate.is_estimate` is true; verify ETA adjustments when mobile cache is enabled.
4. **Test permission boundaries**: Verify `get-cache-status` and `get-cache-health` require `rocket_manage_options` capability.

### Affected Features & Quality Assurance Scope

- **Preload module** (`inc/Engine/Preload/*`): New abilities layer querying the `wp_wpr_rocket_cache` table.
- **MCP/Abilities API integration** (`inc/Engine/Abilities/`, `inc/common/*`): Reuses existing registration patterns and hooks (`wp_abilities_api_init`, `wp_abilities_api_categories_init`).
- **No cache-clearing logic in this PR**: the `clear-cache` ability was removed; cache-clearing abilities are handled entirely by #8623 (merged) and #8598 (open).

### Documentation

**Architecture:** The two abilities follow the established pattern in this codebase (e.g., `RocketInsights\Abilities\*`), placed in `inc/Engine/Preload/Abilities/` and wired through the Preload module's `ServiceProvider`. A shared `Subscriber` class registers the abilities on WordPress core's Abilities API hooks; the `wp-rocket-cache` category itself is registered by `Cache\Abilities\Subscriber` (from `#8623`) and reused here by category slug.

**Two abilities:**
1. **`wp-rocket/get-cache-status`** (read-only, category `wp-rocket-cache`, capability `rocket_manage_options`):
   - Input: exactly one of `url`, `post_id`, or `term_id` + `taxonomy`.
   - Output: `resolved_url`, `tracked` (boolean), `status` (pending/in-progress/completed/failed or null), `modified` (ISO 8601), `last_accessed` (ISO 8601), and a `note` explaining why the URL is not tracked (Preload disabled, excluded, not yet crawled).
   - URL resolution mirrors existing patterns (`get_permalink()`, `get_term_link()`, `rocket_add_url_protocol()`), normalized against `home_url()`'s scheme/host before lookup.

2. **`wp-rocket/get-cache-health`** (read-only, category `wp-rocket-cache`, capability `rocket_manage_options`):
   - Input: none.
   - Output: status counts (pending/in-progress/completed/failed), `tracking_enabled` flag, and an ETA estimate object.
   - ETA formula: `pending_count / batch_size * cron_interval_seconds`, adjusted 2x if mobile cache is active (sequential desktop+mobile fetch per URL halves throughput).
   - Transparently labeled as an estimate with the formula documented in the `method` field.

**Data integrity:**
- Status enum (pending/in-progress/completed/failed) comes directly from the Preload queue — no synthesized or inferred statuses.
- Untracked URLs return `tracked: false` with honest explanations, not fabricated "cached" states.

### New dependencies

None.

### Risks

Low. The implementation leverages battle-tested functions (`Preload\Database\Queries\Cache`) and follows the established Abilities pattern already in production (e.g., `RocketInsights`). Removing the `clear-cache` ability eliminates the overlap risk with #8623/#8598 entirely for this PR.

# Mandatory Checklist

## Code validation

- [x] I validated all the Acceptance Criteria. If possible, provide screenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [x] I implemented built-in tests to cover the new/changed code.

## Code style

- [x] I wrote a self-explanatory code about what it does.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unnecessary complexity.
- [x] Output messages (errors, notices, logs) are explicit enough for users to understand the issue and are actionnable.

## Unticked items justification

*Integration tests for the 2 remaining abilities could not be executed in the implementation sandbox (no bootable WP integration-test environment); they were manually verified line-by-line and independently confirmed live by QA against a running WP 7.0.2 install.*

# Additional Checks
- [ ] In the case of complex code, I wrote comments to explain it.
- [ ] When possible, I prepared ways to observe the implemented system (logs, data, etc.).
- [ ] I added error handling logic when using functions that could throw errors (HTTP/API request, filesystem, etc.)

### CODE REVIEW

File: inc/Engine/Preload/Abilities/PurgeCache.php [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+	 *
+	 * @param string $url URL to check.
+	 *
+	 * @return bool
+	 */
+	private function is_url_in_scope( string $url ): bool {
+		$home = untrailingslashit( home_url() );
+		$url  = untrailingslashit( $url );
+
+		return 0 === strpos( $url, $home );
```
Comment (wordpressfan) [inline inc/Engine/Preload/Abilities/PurgeCache.php]: [HIGH] `is_url_in_scope()` is a naive `strpos` prefix check with no boundary, so it does not correctly implement the required "path-scoped against home_url()" fix.

Confirmed with a plain PHP repro:
```php
var_dump(strpos('https://example.com.attacker.net/page', 'https://example.com') === 0); // true — cross-domain URL wrongly accepted
var_dump(strpos('https://example.com/site10/page', 'https://example.com/site1') === 0); // true — sibling subsite still wrongly accepted
```
Because there is no check that the character following `$home` is either absent or a `/`: (1) the cross-domain guard from AC #5 can be bypassed by any URL whose host has `home_url()`'s host as a literal string prefix (e.g. `example.com.attacker.net`), and (2) the subdirectory-multisite sibling bug this fix was supposed to close still reproduces whenever one site slug is a string-prefix of another (`site1` vs `site10`). Neither existing test (`testShouldRejectUrlOnDifferentHost`, `testShouldRejectUrlOutsideSubdirectorySitePath`) exercises this boundary, which is why it slipped through.

**Fix:** require an exact match or a `/` boundary right after `$home`.

```suggestion
	private function is_url_in_scope( string $url ): bool {
		$home = untrailingslashit( home_url() );
		$url  = untrailingslashit( $url );

		return $url === $home || 0 === strpos( $url, $home . '/' );
	}
```

**Comment by hellofromahmed** [comment]
> [!NOTE]
> Generated by the AI delivery pipeline (lead-reviewer · Claude Sonnet 5).

**Review: ✅ PASS**

Both HIGH blockers from the previous pass are resolved in `90c31b36e`:
- `inc/Engine/Preload/Abilities/PurgeCache.php:284` — `is_url_in_scope()` now does exact-match-or-`/`-bounded-prefix (`$url === $home || 0 === strpos( $url, $home . '/' )`). Verified this closes both the `example.com.attacker.net` cross-domain bypass and the `/site1` vs `/site10` multisite-sibling bypass; both are now covered by new unit tests.
- `inc/Plugin.php` — branch rebased cleanly onto current `develop`; `cache_abilities_subscriber` (#8623) is preserved alongside the new `preload_abilities_subscriber` in `init_abilities_subscribers()`. No duplicate `wp_register_ability_category('wp-rocket-cache', ...)` remains — only `Cache\Abilities\Subscriber` (develop) registers that category now; our `Preload\Abilities\Subscriber` only declares `'category' => 'wp-rocket-cache'` on its own abilities, unchanged. PR is `MERGEABLE`, no conflicts.

Rest of the data contract, edge cases, and tests from the original review remain intact (fix commit was scoped only to these two issues plus their regression tests).

**Nice-to-haves (unchanged, non-blocking):**
- `inc/Engine/Preload/Abilities/CheckCacheStatus.php` — `resolved_url` in the response is the pre-normalization URL, while the DB lookup uses the scheme/host-normalized version; consider returning the normalized value (or documenting the distinction).
- `inc/Engine/Preload/Abilities/CheckCacheStatus.php::normalize_url_scheme_and_host()` — host comparison after stripping `www.` is case-sensitive; consider `strtolower()` on both sides.

**Comment by hellofromahmed** [comment]
> [!NOTE]
> Generated by the AI delivery pipeline (qa-engineer · claude-sonnet-5).

**QA: ✅ PASS**

Boot: `gh pr checkout 8665` + `bin/dev-up.sh` succeeded (exit 0), `http://localhost:8888` reachable (WP 7.0.2, Abilities API present). Validated via Strategy A (WP-CLI `wp eval` against `wp_get_ability(...)->execute(...)`) + Strategy C (existing unit/integration test suite, 30/30 passing for the new Preload Abilities classes).

| Acceptance Criterion | Method | Result |
|---|---|---|
| 1. get-cache-status: honest tracked/untracked, status enum, ISO8601 | API | ✅ Live: seeded row returned `tracked:true, status:completed, modified:2026-07-31T10:00:00Z`; untracked URL returned `tracked:false` + explanatory note; multi-identifier and term-without-taxonomy validation errors returned correctly |
| 2. get-cache-health: counts, tracking_enabled, ETA is_estimate:true | API | ✅ Live: counts matched DB (`completed:1`, others 0), `tracking_enabled:true`, `estimate.is_estimate:true`, `method` field present |
| 3. clear-cache: url/post/domain scopes accepted:true | API + Analysis | ✅ url/domain verified live (`accepted:true`); post-scope live call is blocked locally by the pre-existing `rocket_valid_key()` license guard (`inc/main.php:97`, gates loading of `rocket_clean_post()`) — verified instead via passing unit (`testShouldClearPostScope`) and integration (`testShouldClearPostScopeEndToEnd`) tests |
| 4. cloudflare_purge_triggered mapping (true/false/null) | Analysis (unit tests) | ✅ All scope × do_cloudflare combinations covered and passing in `PurgeCache/ExecuteTest.php`; live re-confirmation was inconclusive due to `Options_Data` option caching not reflecting `update_option()` mid-request (pre-existing options-caching behavior, unrelated to this PR's logic) |
| 5. clear-cache rejects cross-domain URL | API | ✅ Live: `https://evil.example.com/...` rejected with explicit `error`, `accepted:false`, no cache function invoked |
| 6. clear-cache scope=url + homepage uses rocket_clean_home() | API + Code review | ✅ Live: homepage URL cleared successfully; code confirms dedicated `is_homepage_url()` branch calling `rocket_clean_home()` instead of `rocket_clean_files()`, matching admin-bar precedent |
| 7. clear-cache succeeds with no precondition | API + Analysis | ✅ Live + `testShouldSucceedEvenWhenNoProblemDetected` passing — no gating check before calling `rocket_clean_*()` |
| 8. Capability enforcement + WP<6.9 no-op | API + Code review | ✅ Live: subscriber user denied both `get-cache-health` (needs `rocket_manage_options`) and `clear-cache` (needs `rocket_purge_cache`); WP<6.9 no-op verified via code (`function_exists('wp_register_ability')` guards in every `register()`) and integration test skip-guards, not live-tested against an actual pre-6.9 core |

**Not verifiable locally (flagged by the spec itself as requiring live environments):**
- Cloudflare end-to-end purge confirmation on a real Cloudflare-connected site
- Multisite (subdirectory/subdomain/domain-mapped) site-scoping confirmation
- Multilingual (WPML/Polylang) URL resolution and `lang` passthrough confirmation

**Recommendation (non-blocking):** Once merged onto current `develop`, this PR's `wp-rocket-cache` ability category slug collides with the already-merged `Cache\Abilities\Subscriber` (from #8623), which registers the same category slug. This produces a harmless-but-visible `WP_Ability_Categories_Registry::register` "already registered" PHP notice; abilities themselves still register and function correctly. Recommend a unique category slug or reusing the existing category registration.

File: inc/Engine/Preload/Abilities/PurgeCache.php (hunk 1 above)
Comment (wordpressfan) [reply inc/Engine/Preload/Abilities/PurgeCache.php]: Fixed in 90c31b36e — `is_url_in_scope()` now requires an exact match or a `/`-bounded prefix (`$url === $home || 0 === strpos( $url, $home . '/' )`), which closes both the cross-domain (`example.com.attacker.net`) and subdirectory-multisite-sibling (`/site1` vs `/site10`) bypasses. Verified with the repro cases plus the two new unit tests added in that commit (`testShouldRejectUrlWithHostThatIsStringPrefixOfHomeHost`, `testShouldRejectSiblingSubdirectorySiteWithNumericPrefixCollision`).

**Comment by hellofromahmed** [comment]
> [!NOTE]
> Generated by the AI delivery pipeline (qa-engineer · claude-sonnet-5) — follow-up re-verification of commit `90c31b36e`, scoped only to the two items it changed. Original full QA pass: #issuecomment-5141983780 (PASS on all 8 criteria).

**QA (re-verification): ✅ PASS**

Environment: same branch checked out via `gh pr checkout 8665`, `bin/dev-up.sh` succeeded (exit 0), WP 7.0.2 at `http://localhost:8888`, single-site.

| Item re-verified | Method | Result |
|---|---|---|
| AC5 boundary case: `is_url_in_scope()` no longer allows string-prefix host bypass | API (`wp eval` → `wp_get_ability('wp-rocket/clear-cache')->execute()`) | ✅ `http://localhost:8888.evil.net/page` → `accepted:false`, explicit error, `rocket_clean_files()` not invoked. Legit `http://localhost:8888/normal-page/` still `accepted:true`. |
| AC5 boundary case: sibling subdirectory-multisite collision (`/site1` vs `/site10`) | API (`wp eval`, `home_url` filtered to simulate a subdirectory install) | ✅ `http://localhost:8888/site10/page` against home `.../site1` → `accepted:false`. Legit `.../site1/normal-page/` still `accepted:true`. |
| "wp-rocket-cache already registered" notice no longer fires | Analysis (inspected `$wp_filter['wp_abilities_api_categories_init']` callback list) + structural code check | ✅ Only one callback remains hooked to `wp_abilities_api_categories_init` for this category: `Cache\Abilities\Subscriber::register_cache_category`. `Preload\Abilities\Subscriber::get_subscribed_events()` no longer has an entry for that hook at all — the redundant registration is fully removed. |
| Updated unit tests | Test suite | ✅ `composer test-unit -- --filter=ExecuteTest tests/Unit/inc/Engine/Preload/Abilities/PurgeCache/ExecuteTest.php` → 17/17 passing, including the two new boundary-case tests (`testShouldRejectUrlWithHostThatIsStringPrefixOfHomeHost`, `testShouldRejectSiblingSubdirectorySiteWithNumericPrefixCollision`). |

AC1–4, 6–8 were not re-tested — unaffected by this fix, already PASS in the original report.

**Comment by hellofromahmed** [comment]
Removed the `wp-rocket/clear-cache` ability from this PR (commit 34e6d83). It duplicated cache-clearing functionality already covered elsewhere in the same `wp-rocket-cache` category:

- `scope=domain` reimplemented #8623's `wp-rocket/clear-website-cache` (already merged to develop) — both just call `rocket_clean_domain()`.
- `scope=url` reimplemented #8598's `wp-rocket/clear-url-cache` (open) — same site-ownership validation, homepage special-case via `rocket_clean_home()`, and `rocket_clean_files()` call.

This PR now only adds the two read-only abilities, `get-cache-status` and `get-cache-health`. Removed the ability file, its Subscriber/ServiceProvider wiring, and its unit/integration tests; updated the PR description accordingly. Post-level cache clearing via MCP can be proposed in a separate follow-up if still wanted.

File: inc/Engine/Preload/Abilities/CheckCacheStatus.php [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+
+	/**
+	 * Executes the ability to get the Preload cache status for a URL, post, or term.
+	 *
+	 * @param array|null $input Input containing url, or post_id, or term_id + taxonomy.
+	 *
+	 * @return array
+	 */
+	public function execute( $input = null ): array {
+		$this->track_event(
```
Comment (Khadreal) [inline inc/Engine/Preload/Abilities/CheckCacheStatus.php:140]: Considering that this execution could fail, shouldn't we track only when it didn't fail, so maybe move this track event a bit down

File: tests/Integration/inc/Engine/Preload/Abilities/CheckCacheHealth/RegisterCheckCacheHealthAbilityTest.php [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+		$ability = wp_get_ability( self::ABILITY_ID );
+		$this->assertNotNull( $ability );
+
+		$result = $ability->execute();
+
+		$this->assertInstanceOf( 'WP_Error', $result, 'Should return WP_Error when user lacks rocket_manage_options.' );
+	}
+
+	public function testShouldReturnCountsMatchingSeededRows(): void {
```
Comment (Khadreal) [inline tests/Integration/inc/Engine/Preload/Abilities/CheckCacheHealth/RegisterCheckCacheHealthAbilityTest.php:77]: shouldn't we move the test data to fixtures directory for consistency

**Review by Khadreal — COMMENTED** [review]
Thanks @wordpressfan. LGTM except the question I have about the placement of the tracking and moving the test data to fixtures directory

File: inc/Engine/Preload/Abilities/CheckCacheStatus.php (hunk 2 above)
Comment (wordpressfan) [reply inc/Engine/Preload/Abilities/CheckCacheStatus.php:140]: Checked this against the rest of the codebase before changing it: every other MCP ability (`ClearUrlCache::execute()`, `ClearWebsiteCache::execute()`, `SetOption::execute()`, `GetOptions::execute()`, and `CheckCacheHealth::execute()` right next to this one) fires `track_event()` unconditionally as the first line, before any validation or work happens. That looks like a deliberate convention: "Ability Executed" tracks that the ability was invoked, not that it succeeded.

If we move it past validation here, this ability's event stops meaning the same thing as every other ability's event — invocation counts would no longer be comparable across abilities, and failed/invalid calls (which are still useful signal, e.g. malformed MCP client requests) would go untracked here but not elsewhere.

If the intent is actually to track success-only, I think that should be a deliberate change applied consistently across all abilities rather than just this one — happy to open that as a follow-up if there's appetite for it, but I'd rather not special-case this one ability. Let me know if I'm missing context on why this one should differ.

File: tests/Integration/inc/Engine/Preload/Abilities/CheckCacheHealth/RegisterCheckCacheHealthAbilityTest.php (hunk 3 above)
Comment (wordpressfan) [reply tests/Integration/inc/Engine/Preload/Abilities/CheckCacheHealth/RegisterCheckCacheHealthAbilityTest.php:77]: Applied in 28b8f13a0 — moved the seeded rows to `tests/Fixtures/inc/Engine/Preload/Abilities/CheckCacheHealth/RegisterCheckCacheHealthAbilityTest.php`, keyed by test method name and loaded via `WP_ROCKET_TESTS_FIXTURES_DIR` (the same constant `tests/Integration/bootstrap.php` already uses for plain-array fixtures). Both `testShouldReturnCountsMatchingSeededRows` and `testShouldReturnNullEstimateWhenTrackingDisabled` now pull their seed rows from there instead of inlining them.

**Comment by Mai-Saad** [comment]
@wordpressfan Thanks for the PR, Working fine except for the following:
- If we disable mobile cache using filter, then clear and preload then check health status in Claude will mention time as if mobile cache is on, do we need change for that? => fixed on latest commit :white_check_mark: 
- If we have URL with non-latin char, preload of original URL fails while another entry encoded i.e %0d ..etc added in DB and that is completed and is the one AI check  {preload behavior is already on 3.23.1.1 independent on MCP , may be we need GH for that} Related thread https://group-onecom.slack.com/archives/C08F4LZB2QG/p1785849502487559?thread_ts=1785847972.719509&cid=C08F4LZB2QG

Todo:

- [x] explore on multisite => tested on subdirectory setup while basic auth , as oauth not compatible with multisite (note, due to issue with transients on multisite sometimes not deleted, we may see global score on subsite also no URL in RI list of this site {probably it was there before reinstall})
- [x] explore on multilingual => check on subdirectory setup, subdomain env not available so out of scope, qs {we now don't add URL with qs to cache table so URL won't be tracked when get cache status}
- [x] explore with CF Cache
